Fixing Travis CI and use go 1.10 and later versions

xinfengliu opened the pull request:
- Need to upgrade go 1.10 and later to fix "go get -u github.com/golang/lint/golint" error in travis CI.
See: https://github.com/golang/lint/issues/421

- Also modify integration test code to make it pass the CI when using golang 1.10.x.

- Replace "github.com/golang/lint/golint" with "golang.org/x/lint/golint" (This is needed for golang 1.12 or later)

- Modify `Makefile` to fix `make test` error when using go 1.12 or later.

Signed-off-by: Xinfeng Liu <xinfeng.liu@gmail.com>

xinfengliu committed:
Update travis to use go 1.10.x for fixing CI

To fix "go get -u github.com/golang/lint/golint" error.
https://github.com/golang/lint/issues/421

Also fix integratinon test when using golang 1.10.x
In AddProcessToContainer(), the returned stdout buffer has x00 padded, after it is converted to string, it still has padded zeros.
This commit trims the padded zeros in string.

Replace "github.com/golang/lint/golint" with "golang.org/x/lint/golint" from below error message:
(this is needed for golang 1.12 or later)
```
package github.com/golang/lint/golint: code in directory /home/travis/gopath/src/github.com/golang/lint/golint expects import "golang.org/x/lint/golint"
```

Signed-off-by: Xinfeng Liu <xinfeng.liu@gmail.com>

thaJeztah commented on integration-test/exec_test.go:
```diff
@@ -58,7 +59,7 @@ func (cs *ContainerdSuite) TestBusyboxTopExecEcho(t *check.C) {
 	}
 
-	t.Assert(echop.io.stdoutBuffer.String(), checker.Equals, "Ay Caramba!")
+	t.Assert(strings.TrimRight(echop.io.stdoutBuffer.String(), string('\x00')), checker.Equals, "Ay Caramba!")
```
Was this due to a change in Go 1.10?

xinfengliu replied:
Yes.

thaJeztah commented on .travis.yml:
```diff
@@ -4,8 +4,7 @@ sudo: required
 go:
-  - 1.8.x
-  - tip
+  - 1.10.x
```
I think docker 17.06 ee is currently built with Go 1.12.x, so it may be good to have that version in this list at least (but given that Go 1.12.x will reach EOL soon, probably have 1.13.x in there as well).

Any reason to remove `tip`?

xinfengliu replied:
OK, I will add go 1.12, 1.13. `tip` also caused golint error.

thaJeztah commented:
could you either squash the first two commits, or swap their order? If the second commit is needed to fix CI for changes due to updating to Go 1.10, it should either be there first commit, or go together with the actual change to update to Go 1.10 (otherwise Git bisect would be broken)

xinfengliu commented:
@thaJeztah , thanks. I squshed the commits.
However, I could not make go 1.12 or 1.13 work, `make test` failed at ` go test -bench=. ` :
```
FAIL	github.com/containerd/containerd/runtime [build failed]
FAIL	github.com/containerd/containerd/supervisor [build failed]
```
Without `-bench=.`, `make test` could pass. I don't know why.

xinfengliu committed:
Add go 1.12, 1.13 and tip to Travis CI

Also fixed `make test` failure in go 1.12 and 1.13 by removing 'github.com/containerd/containerd/api/http/pprof' from 'go test -bench'.
Set parallel to 1 in `make integration-test`

Signed-off-by: Xinfeng Liu <xinfeng.liu@gmail.com>

xinfengliu commented:
The culprit is `github.com/containerd/containerd/api/http/pprof`. There is no test or benchmark method in that package, so it's safe to take this package out of `go test`.

thaJeztah commented:
Two failures on `tip`. First one is interesting. (not a blocker, because it's `tip`, but curious what changed).

I kicked CI once more to see if it was just an intermittent failure

```
FAIL: start_linux_test.go:522: ContainerdSuite.TestSigkillShimReuseName
855
856start_linux_test.go:556:
857    t.Assert(*e, checker.Equals, evt)
858... obtained types.Event = types.Event{Type:"exit", Id:"top", Status:0xff, Pid:"init", Timestamp:(*timestamp.Timestamp)(0xc0002ded90)}
859... expected types.Event = types.Event{Type:"exit", Id:"top", Status:0x89, Pid:"init", Timestamp:(*timestamp.Timestamp)(0xc0002ded90)}
860
861
862----------------------------------------------------------------------
863FAIL: start_linux_test.go:569: ContainerdSuite.TestSigkillShimWhileContainerIsPaused
864
865start_linux_test.go:577:
866    t.Fatal(err)
867... Error: rpc error: code = 2 desc = oci runtime error: container with id exists: top
```

thaJeztah reviewed (approved):
LGTM

kicked CI again, but I guess we can merge after that completes (and ignore tip for now)
